mimium-lang/src/compiler/parser.rs

```diff
@@ -46,351 +48,412 @@ fn type_parser() -> impl Parser<Token, Type, Error = Simple<Token>> + Clone {
+        .map(|(id, ty)| TypedId { id, ty })
+        .labelled("lvar")
+}
+
+fn comment_parser() -> impl Parser<Token, (), Error = Simple<Token>> + Clone {
+    select! {Token::Comment(Comment::SingleLine(_))=>(),
+    Token::Comment(Comment::MultiLine(_))=>()}
+}
+
+fn eol() -> impl Parser<Token, Token, Error = Simple<Token>> + Clone {
+    just(Token::LineBreak)
+        .or(just(Token::SemiColon))
+        .or(just(Token::EndOfInput))
+}
+
+fn stmt_parser() -> impl Parser<Token, WithMeta<Expr>, Error = Simple<Token>> + Clone {
+    recursive(|stmt| {
+        // Note: this needs to be declare() first instead of recursive()
+        // directly because otherwise the dependency cannot be described.
+        let mut expr = Recursive::<Token, WithMeta<Expr>, Simple<Token>>::declare();
+        let mut non_block_expr = Recursive::<Token, WithMeta<Expr>, Simple<Token>>::declare();
+        let mut block_expr = Recursive::<Token, WithMeta<Expr>, Simple<Token>>::declare();
+
+        expr.define(non_block_expr.clone().or(block_expr.clone()));
+        non_block_expr.define(non_block_expr_parser(expr.clone()));
+        block_expr.define(block_expr_parser(expr.clone()));
+
+        let expr_stmt = expr_stmt_parser(expr.clone(), stmt.clone());
+        let let_stmt = let_stmt_parser(expr.clone(), stmt.clone());
+        let func_decl = func_decl_parser(block_expr.clone(), stmt.clone());
+        let macro_decl = macro_decl_parser(block_expr.clone(), expr.clone());
+
+        expr_stmt.or(let_stmt).or(func_decl).or(macro_decl)
+    })
+}
+
+fn op_parser<'a, P>(
+    expr: P,
+    ops: Vec<Op>,
+) -> impl Parser<Token, WithMeta<Expr>, Error = Simple<Token>> + Clone + 'a
+where
+    P: Parser<Token, WithMeta<Expr>, Error = Simple<Token>> + Clone + 'a,
+{
+    let op_cls = |x: WithMeta<_>, y: WithMeta<_>, op: Op, opspan: Span| {
+        WithMeta(
+            Expr::Apply(
+                Box::new(WithMeta(
+                    Expr::Var(op.get_associated_fn_name().to_string(), None),
+                    opspan,
+                )),
+                vec![x.clone(), y.clone()],
+            ),
+            x.1.start..y.1.end,
+        )
+    };
+    let optoken = move |o: Op| {
+        just(Token::Op(o)).map_with_span(|e, s| {
+            (
+                match e {
+                    Token::Op(o) => o,
+                    _ => Op::Unknown(String::from("invalid")),
+                },
+                s,
+            )
+        })
+    };
+
+    let lhs = expr.clone();
+    let rhs = expr;
+
+    let ops: Vec<_> = ops.into_iter().map(optoken).collect();
+
+    lhs.then(choice(ops).then(rhs).repeated())
+        .foldl(move |x, ((op, opspan), y)| op_cls(x, y, op, opspan))
+        .boxed()
+}
+
+fn non_block_expr_parser<'a>(
+    expr: impl Parser<Token, WithMeta<Expr>, Error = Simple<Token>> + Clone + 'a,
+) -> impl Parser<Token, WithMeta<Expr>, Error = Simple<Token>> + Clone + 'a {
```

Comment:
By taking a argument of `impl Parser`, this function can take either 

1. a recursive parser passed from `recursive()`
2. a non-recursive parser like `literal_expr_parser()`

Because of 1., we can separate functions instead of putting all the code into the lambda of `recursive()`. By doing this, probably parser test can be written more easily.